hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/SparkHoodieTableFileIndex.scala

```diff
@@ -187,12 +189,16 @@ class SparkHoodieTableFileIndex(spark: SparkSession,
    * @param predicates     The filter condition.
    * @return The pruned partition paths.
    */
-  protected def prunePartition(partitionPaths: Seq[PartitionPath], predicates: Seq[Expression]): Seq[PartitionPath] = {
+  protected def prunePartition(predicates: Seq[Expression]): Seq[PartitionPath] = {
     val partitionColumnNames = partitionSchema.fields.map(_.name).toSet
     val partitionPruningPredicates = predicates.filter {
       _.references.map(_.name).toSet.subsetOf(partitionColumnNames)
     }
     if (partitionPruningPredicates.nonEmpty) {
+      val equalPredicate = partitionPruningPredicates.filter(_.isInstanceOf[EqualTo])
+      val names = equalPredicate.map(_.asInstanceOf[EqualTo]).map(_.left.asInstanceOf[AttributeReference].name).toArray
```

Comment:
We can not assume what those predicates might be

Reply:
Yes, but here the lazy file fetching is only possible for equal predicate (maybe also `in` predicate). The code is kind of hard-coded. Any better way to achieve this would be appreciated! Not very familiar with spark scala coding.

Reply:
Yeah, we need to make it a proper restriction on the semantic:

 - When filters are passed in we check whether we could do lazy-merging on them (checking if its strictly `EqualTo`, `In` predicates) 
 - If not, we fallback to eager loading

Reply:
There is also another case, where we eager loading all partition paths and fetch file group lazily. E.g., User specifies a complex predicates on partitions or only gives a single predicate for the second partition field. 

So maybe restrict to the following two cases:
- We have predicate on partition columns, we do lazy loading of file groups. And partition loading based on prefix is another optimization to avoid full partition listing.
- If not, we fallback to eager loading of partitions and file groups.

By the way, currently I put an option to control the loading behavior, `REFRESH_PARTITION_AND_FILES_IN_INITIALIZATION`. With the above semantic, maybe we can remote it and let it be handled automatically. WDYT?

Reply:
> By the way, currently I put an option to control the loading behavior, REFRESH_PARTITION_AND_FILES_IN_INITIALIZATION. With the above semantic, maybe we can remote it and let it be handled automatically. WDYT?

Sorry, not sure i fully follow this one: what do you want to remove?

Reply:
Ah never mind about this one. I was thinking to remove the config, and handle the lazy listing internally (i.e., turn on or not).